unicodetools/src/main/java/org/unicode/text/UCA/WriteCharts.java

```diff
@@ -909,46 +899,52 @@ static short getBestScript(String s) {
     // static final UnicodeMap<String> SCRIPT_EXTENSIONS =
     // INDEX_UNICODE_PROPS.load(UcdProperty.Script_Extensions);
 
-    static BitSet getBestScript(int original, String transformed, BitSet toReturn) {
-        toReturn.clear();
-        addScript(original, toReturn);
-        if (transformed != null) {
-            int cp;
-            for (int i = 0; i < transformed.length(); i += UTF16.getCharCount(cp)) {
-                cp = UTF16.charAt(transformed, i);
-                addScript(cp, toReturn);
+    /**
+     * Returns the best explicit Script value for cp or else for decomp. If there is no such
+     * explicit script, returns a negative value. Never returns Common or Inherited.
+     */
+    private static int getBestScript(int cp, String decomp) {
+        int sc = getExplicitScript(cp);
+        if (sc < 0 && !equals(decomp, cp)) {
+            for (int i = 0; sc < 0 && i < decomp.length(); i += Character.charCount(cp)) {
+                cp = decomp.codePointAt(i);
+                sc = getExplicitScript(cp);
```

Comment:
Last one wins…

Reply:
No, the "for" condition includes `sc < 0` which means when we get the first explicit script we stop.

Reply:
Ah, there it is! I think it would be cleaner to explicitly check and break after calling the function, but I’m not going to argue with working code.

Reply:
I like it concise like this...

Reply:
I also thought about using the IntStream functionality, but it might take some playing with the syntax and I am not sure that it would be any more readable, or any faster. (More object creation vs. better compiler optimizations.) Something like
```
decomp.codePoints().map(cp to script).filter(non-negative).findFirst().ifPresent(set sc);
```